### ISSUE TITLE
2.2.9 Concurrency does not seem to work properly

### ISSUE THREAD

**Issue by Mertsch** [microsoft/testfx#1063]
## Description
I am getting various error messages when running multiple test in parallel. When tests are executes one by one, no problem.

## Steps to reproduce
Run .NET 6.0 tests with `[assembly: Parallelize(Scope = ExecutionScope.ClassLevel)]`

## Expected behavior
No exceptions

## Actual behavior
```
Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)
```
```
Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)
```
```
Exception thrown while executing test. If using extension of TestMethodAttribute then please contact vendor. Error message: Index was outside the bounds of the array., Stack trace:    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)
   at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.LogMessageListener..ctor(Boolean captureDebugTraces)
   at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestMethodInfo.Invoke(Object[] arguments)
   at Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute.Execute(ITestMethod testMethod)
   at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestMethodRunner.ExecuteTest(TestMethodInfo testMethodInfo)
Stacktrace is not available
```

## Environment
ReSharper TestRunner
```
2022.04.08 20:34:27.796   ERROR An item with the same key has already been added. Key: context, reason: System.ArgumentException: An item with the same key has already been added. Key: context at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior) at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value) at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder() at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType) at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.UnitTestRunner.RunSingleTest(TestMethod testMethod, IDictionary`2 testContextProperties) at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.ExecuteTestsWithTestRunner(IEnumerable`1 tests, IRunContext runContext, ITestExecutionRecorder testExecutionRecorder, String source, IDictionary`2 sourceLevelParameters, UnitTestRunner testRunner) at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.<>c__DisplayClass17_1.<ExecuteTestsInSource>b__6() at System.Threading.Tasks.Task.InnerInvoke() at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj) at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) --- End of stack trace from previous location --- at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)

--- EXCEPTION #1/2 [RdFault]
Message = “
  An item with the same key has already been added. Key: context, reason: System.ArgumentException: An item with the same key has already been added. Key: context
     at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
     at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
     at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
     at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.UnitTestRunner.RunSingleTest(TestMethod testMethod, IDictionary`2 testContextProperties)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.ExecuteTestsWithTestRunner(IEnumerable`1 tests, IRunContext runContext, ITestExecutionRecorder testExecutionRecorder, String source, IDictionary`2 sourceLevelParameters, UnitTestRunner testRunner)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.<>c__DisplayClass17_1.<ExecuteTestsInSource>b__6()
     at System.Threading.Tasks.Task.InnerInvoke()
     at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
  --- End of stack trace from previous location ---
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
”
ExceptionPath = Root.InnerException
ClassName = JetBrains.Rd.Tasks.RdFault
Data.ThreadLocalDebugInfo = “call `ServerSocket-<noname>.JetBrains.ReSharper.TestRunner.Abstractions.IClientEndpoint.SendMessage_proxy` (15325296026731086546)”
HResult = COR_E_EXCEPTION=80131500
Source = mscorlib
StackTraceString = “
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.Activation.ServerMessageBroker.<SendMessage>d__2.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.<<RunTests>b__8_1>d.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<CancelIf>d__15.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<ToEmptyTask>d__13.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<ThrowIf>d__14`2.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.BoundTestRunnerAgent.<RunSafe>d__11.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.<RunTests>d__8.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.<Run>d__16.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.Launch.UnitTestRun.<StartCore>d__37.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at JetBrains.Util.Special.GeneralUtil.WithNotNull[T](T item, Action`1 F)
     at JetBrains.ReSharper.UnitTestFramework.Execution.Launch.UnitTestRun.<StartCore>d__37.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at JetBrains.ReSharper.UnitTestFramework.Execution.Launch.UnitTestRun.<Start>d__36.MoveNext()
”

--- Outer ---

--- EXCEPTION #2/2 [LoggerException]
Message = “
  An item with the same key has already been added. Key: context, reason: System.ArgumentException: An item with the same key has already been added. Key: context
     at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
     at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
     at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
     at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.UnitTestRunner.RunSingleTest(TestMethod testMethod, IDictionary`2 testContextProperties)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.ExecuteTestsWithTestRunner(IEnumerable`1 tests, IRunContext runContext, ITestExecutionRecorder testExecutionRecorder, String source, IDictionary`2 sourceLevelParameters, UnitTestRunner testRunner)
     at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.<>c__DisplayClass17_1.<ExecuteTestsInSource>b__6()
     at System.Threading.Tasks.Task.InnerInvoke()
     at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
  --- End of stack trace from previous location ---
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
”
ExceptionPath = Root
ClassName = JetBrains.Util.LoggerException
InnerException = “Exception #1 at Root.InnerException”
HResult = COR_E_APPLICATION=80131600
StackTraceString = “
  at JetBrains.Util.ILoggerEx.LogException(ILogger this, LoggingLevel level, Exception exception, ExceptionOrigin exceptionOrigin, String comment)
     at JetBrains.Util.ILoggerEx.Error(ILogger this, Exception exception, ExceptionOrigin origin, String comment)
     at JetBrains.ReSharper.UnitTestFramework.Execution.Launch.UnitTestRun.<Start>d__36.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.Launch.UnitTestRun.<StartCore>d__37.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.<Run>d__16.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.<RunTests>d__8.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.BoundTestRunnerAgent.<RunSafe>d__11.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<ThrowIf>d__14`2.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at System.Threading.Tasks.TaskFactory.CompleteOnInvokePromise.Invoke(Task completingTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<ToEmptyTask>d__13.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<CancelIf>d__15.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at System.Threading.Tasks.TaskFactory.CompleteOnInvokePromise.Invoke(Task completingTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.<<RunTests>b__8_1>d.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.Activation.ServerMessageBroker.<SendMessage>d__2.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
     at System.Threading.Tasks.Task.FinishContinuations()
     at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
     at System.Threading.Tasks.Task`1.TrySetException(Object exceptionObject)
     at System.Threading.Tasks.TaskCompletionSource`1.TrySetException(Exception exception)
     at System.Threading.Tasks.TaskCompletionSource`1.SetException(Exception exception)
     at JetBrains.Collections.Viewable.SignalBase`1.Fire(T value)
     at JetBrains.Collections.Viewable.WriteOnceProperty`1.WriteOnceSignal.Fire(T value)
     at JetBrains.Collections.Viewable.WriteOnceProperty`1.SetIfEmpty(T newValue)
     at JetBrains.Rd.Tasks.WiredRdTask`2.CallSite.OnWireReceived(UnsafeReader reader)
     at JetBrains.Rd.Impl.MessageBroker.Execute(IRdWireable reactive, Byte[] msg)
     at JetBrains.Rd.Impl.MessageBroker.Dispatch(RdId id, Byte[] msg)
     at JetBrains.Diagnostics.LogEx.Catch(ILog log, Action action)
     at JetBrains.Rd.WireBase.Receive(Byte[] msg)
     at JetBrains.Rd.Impl.SocketWire.Base.ReadMsg()
     at JetBrains.Rd.Impl.SocketWire.Base.ReceiverProc(Socket socket)
     at JetBrains.Rd.Impl.SocketWire.Base.<>c__DisplayClass19_0.<.ctor>b__1(Socket socket)
     at JetBrains.Collections.Viewable.SignalBase`1.Fire(T value)
     at JetBrains.Collections.Viewable.ViewableProperty`1.set_Value(T value)
     at JetBrains.Rd.Impl.SocketWire.Server.<>c__DisplayClass4_0.<StartServerSocket>b__1()
     at JetBrains.Diagnostics.LogEx.Catch(ILog log, Action action)
     at JetBrains.Rd.Impl.SocketWire.Server.<>c__DisplayClass4_0.<StartServerSocket>b__0()
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
     at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.ThreadHelper.ThreadStart()
”
```

**Comment by JustArchi** [comment]
I confirm that version 2.2.9 has concurrency issues. See https://github.com/JustArchiNET/ArchiSteamFarm/pull/2549

**Comment by majastrz** [comment]
Same issue in https://github.com/Azure/bicep/pull/6466.

**Comment by Haplois** [comment]
We are sorry for the issue, investigating the issue.

**Comment by ChristoWolf** [comment]
I am also observing the issue.

**Comment by nohwnd** [comment]
Does anyone have some smaller project that reproduces this as well so I can try? Please :)

**Comment by JustArchi** [comment]
@nohwnd I don't know if mine qualifies as "smaller project", but it should be possible to reproduce easily - https://github.com/JustArchiNET/ArchiSteamFarm

After `git clone`, edit `Direcory.Packages.props` to use `2.2.9` version. Afterwards you can just `dotnet test`, should be easy to reproduce, but might require a few tries as the issue is race condition.

See https://github.com/JustArchiNET/ArchiSteamFarm/pull/2549 for more details.

**Comment by JustArchi** [comment]
@nohwnd Here is example CI run, so you know what you're looking for: https://github.com/JustArchiNET/ArchiSteamFarm/runs/5967640130?check_suite_focus=true

```
Run dotnet test ArchiSteamFarm.Tests -c "Debug" -p:ContinuousIntegrationBuild=true -p:UseAppHost=false --nologo
  Determining projects to restore...
  The project D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.csproj is using CentralPackageVersionManagement, a NuGet preview feature.
  The project D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.Tests\ArchiSteamFarm.Tests.csproj is using CentralPackageVersionManagement, a NuGet preview feature.
  Restored D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.csproj (in 1.56 sec).
  Restored D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.Tests\ArchiSteamFarm.Tests.csproj (in 1.82 sec).
  ArchiSteamFarm -> D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm\bin\Debug\net6.0\ArchiSteamFarm.dll
  ArchiSteamFarm.Tests -> D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.Tests\bin\Debug\net6.0\ArchiSteamFarm.Tests.dll
  ArchiSteamFarm -> D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm\bin\Debug\net48\ArchiSteamFarm.exe
  ArchiSteamFarm.Tests -> D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.Tests\bin\Debug\net48\ArchiSteamFarm.Tests.dll
Test run for D:\a\ArchiSteamFarm\ArchiSteamFarm\ArchiSteamFarm.Tests\bin\Debug\net6.0\ArchiSteamFarm.Tests.dll (.NETCoreApp,Version=v6.0)
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
  Failed RepetitiveCharactersWeakenPassphrases
  Error Message:
   An item with the same key has already been added. Key: err
  Stack Trace:
     at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter.GetOrAddStringBuilder()
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadSafeStringWriter..ctor(IFormatProvider formatProvider, String outputType)


Failed!  - Failed:     1, Passed:    71, Skipped:     0, Total:    72, Duration: 348 ms - ArchiSteamFarm.Tests.dll (net6.0)
```

**Comment by nohwnd** [comment]
Merged a fix, we want to release preview of 2.2.10 today, I am waiting till 3PM (2h from now) for another fix. Then I'll make the release.

**Comment by JustArchi** [comment]
Thank you a lot for very swift response @nohwnd :sunglasses:.

**Comment by nohwnd** [comment]
Pre-release is up: https://www.nuget.org/packages/MSTest.TestAdapter/2.2.10-preview-20220414-01

**Comment by abatishchev** [comment]
Hi, I'm here to report this issue. Thanks for fixing it! When the stable version will be released?

**Comment by nohwnd** [comment]
This was released in  2.2.10
